src/nps/userSurvey.ts

```diff
@@ -0,0 +1,249 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from "vscode";
+import * as constants from "../constants/constants";
+import { ReactWebviewPanelController } from "../controllers/reactWebviewController";
+import {
+    UserSurveyReducers,
+    UserSurveyState,
+} from "../sharedInterfaces/userSurvey";
+import * as locConstants from "../constants/locConstants";
+import { sendActionEvent } from "../telemetry/telemetry";
+import {
+    TelemetryActions,
+    TelemetryViews,
+} from "../sharedInterfaces/telemetry";
+
+const PROBABILITY = 0.15;
+const SESSION_COUNT_KEY = "nps/sessionCount";
+const LAST_SESSION_DATE_KEY = "nps/lastSessionDate";
+const SKIP_VERSION_KEY = "nps/skipVersion";
+const IS_CANDIDATE_KEY = "nps/isCandidate";
+
+export class UserSurvey {
+    private static _instance: UserSurvey;
+    private _webviewController: UserSurveyWebviewController;
+    private constructor(private _context: vscode.ExtensionContext) {}
+    public static createInstance(_context: vscode.ExtensionContext): void {
+        UserSurvey._instance = new UserSurvey(_context);
+    }
+    public static getInstance(): UserSurvey {
+        return UserSurvey._instance;
+    }
+
+    public async promptUserForNPSFeedback(): Promise<void> {
+        const globalState = this._context.globalState;
+        const skipVersion = globalState.get(SKIP_VERSION_KEY, "");
+        if (skipVersion) {
+            return;
+        }
+
+        const date = new Date().toDateString();
+        const lastSessionDate = globalState.get(
+            LAST_SESSION_DATE_KEY,
+            new Date(0).toDateString(),
+        );
+
+        if (date === lastSessionDate) {
+            return;
+        }
+
+        const sessionCount = globalState.get(SESSION_COUNT_KEY, 0) + 1;
+        await globalState.update(LAST_SESSION_DATE_KEY, date);
+        await globalState.update(SESSION_COUNT_KEY, sessionCount);
+
+        if (sessionCount < 9) {
```

Comment:
Can you add comments for the logic here?

`// don't prompt for feedback from users until they've had a chance to use the extension`